File: satellites/gameforge-rs/crates/gf-gameforge/src/lafs.rs
use std::collections::HashMap;
use chrono::{DateTime,Utc};
use serde::{Deserialize,Serialize};
use tokio::sync::RwLock;
use gf_core::digest::sha256_hex;
pub const CHUNK_CAP:usize=16384;pub const CHUNK_MAX_BYTES:usize=1<<20;
#[derive(Debug,Clone,Serialize,Deserialize)]pub struct Chunk{pub digest:String,pub bytes:Vec<u8>,pub stored:DateTime<Utc>}
#[derive(Debug,Clone,Serialize,Deserialize)]pub struct Manifest{pub id:String,pub name:String,pub chunks:Vec<String>,pub total_bytes:u64,pub created:DateTime<Utc>}
pub struct Lafs{chunks:RwLock<HashMap<String,Chunk>>,manifests:RwLock<HashMap<String,Manifest>>}
impl Default for Lafs{fn default()->Self{Self::new()}}
impl Lafs{
 pub fn new()->Self{Self{chunks:RwLock::new(HashMap::new()),manifests:RwLock::new(HashMap::new())}}
 pub async fn put_chunk(&self,bytes:&[u8])->Result<String,&'static str>{if bytes.len()>CHUNK_MAX_BYTES{return Err("chunk exceeds 1 MiB")}let digest=sha256_hex(bytes);let mut chunks=self.chunks.write().await;if chunks.contains_key(&digest){return Ok(digest)}if chunks.len()>=CHUNK_CAP{return Err("chunk store at capacity — compact before writing")}let chunk=Chunk{digest:digest.clone(),bytes:bytes.to_vec(),stored:Utc::now()};chunks.insert(digest.clone(),chunk.clone());drop(chunks);if let Ok(db)=gf_core::db::get_db().await{tokio::spawn(async move{let _=db.collection::<Chunk>("lafs_chunks").insert_one(chunk).await;});}Ok(digest)}
 pub async fn get_chunk(&self,d:&str)->Option<Chunk>{self.chunks.read().await.get(d).cloned()}
 pub async fn pin_manifest(&self,name:&str,digests:Vec<String>)->Result<Manifest,String>{let chunks=self.chunks.read().await;let mut total=0u64;for d in &digests{match chunks.get(d){Some(c)=>total+=c.bytes.len() as u64,None=>return Err(format!("unknown chunk {d}"))}}drop(chunks);let m=Manifest{id:uuid::Uuid::new_v4().to_string(),name:name.to_string(),chunks:digests,total_bytes:total,created:Utc::now()};self.manifests.write().await.insert(m.name.clone(),m.clone());Ok(m)}
 pub async fn read_manifest(&self,name:&str)->Option<Vec<u8>>{let m=self.manifests.read().await.get(name)?.clone();let chunks=self.chunks.read().await;let mut out=Vec::with_capacity(m.total_bytes as usize);for d in &m.chunks{let c=chunks.get(d)?;if sha256_hex(&c.bytes)!=*d{return None}out.extend_from_slice(&c.bytes)}Some(out)}
 pub async fn manifests(&self)->Vec<Manifest>{self.manifests.read().await.values().cloned().collect()}
 pub async fn stats(&self)->serde_json::Value{let c=self.chunks.read().await;let bytes:usize=c.values().map(|x|x.bytes.len()).sum();serde_json::json!({"chunks":c.len(),"bytes":bytes,"manifests":self.manifests.read().await.len()})}
}

File: skeleton/ai/forge/lafs.py
"""LAFS — Labyrinthine Append-only File Store.

Port of gameforge-rs ``crates/gf-gameforge/src/lafs.rs``. Content-addressed
chunks under manifests. A chunk is written once, addressed by its SHA-256
digest (F1 fix — a content-addressed store with a non-cryptographic address
is corruptible by anyone who can write two chunks), and never mutated. The
labyrinth has no erasers, only new rooms. Manifests pin chunk order; every
file in the empire has a provenance chain back through the quorum gate.

This is the forge-path chunk store (hex address = SHA-256), **not** the
Bayesian Lever Arch knowledge ledger under ``backend/gameforge/lafs/``.

Optional light integration: callers that already journal via
``skeleton.forge.outbox.MaterialiseOutbox`` may store blueprint payloads
through ``Lafs.put_chunk`` / ``pin_manifest`` and record digests on the
intent document — extend-only; outbox itself is unchanged.
"""

from __future__ import annotations

import hashlib
import threading
import time
import uuid
from dataclasses import dataclass, field
from typing import Any

from skeleton.kernel.errors import SkeletonError

CHUNK_CAP = 16384
CHUNK_MAX_BYTES = 1 << 20  # 1 MiB


class LafsError(SkeletonError):
    """LAFS refuse — oversize chunk, capacity, unknown digest, corruption."""

    code = "FORGE.LAFS"
    http_status = 400


def _sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


@dataclass
class Chunk:
    """One immutable content-addressed blob."""

    digest: str
    bytes: bytes
    stored: float = field(default_factory=time.time)

    def to_dict(self) -> dict[str, Any]:
        return {
            "digest": self.digest,
            "size": len(self.bytes),
            "stored": self.stored,
        }


@dataclass
class Manifest:
    """Ordered pin of chunk digests assembling one logical file."""

    name: str
    chunks: list[str]
    total_bytes: int
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    created: float = field(default_factory=time.time)

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "name": self.name,
            "chunks": list(self.chunks),
            "total_bytes": self.total_bytes,
            "created": self.created,
        }


class Lafs:
    """In-process content-addressed chunk store + manifest pins.

    Thread-safe via a single RLock (mirrors the RS RwLock covering chunk /
    manifesto mutation so observers never see a half-written room).
    """

    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._chunks: dict[str, Chunk] = {}
        self._manifests: dict[str, Manifest] = {}

    def put_chunk(self, data: bytes) -> str:
        """Store a chunk. Idempotent: same bytes, same room. Returns digest."""
        if len(data) > CHUNK_MAX_BYTES:
            raise LafsError("chunk exceeds 1 MiB")
        digest = _sha256_hex(data)
        with self._lock:
            if digest in self._chunks:
                return digest  # already in the labyrinth
            if len(self._chunks) >= CHUNK_CAP:
                raise LafsError("chunk store at capacity — compact before writing")
            self._chunks[digest] = Chunk(digest=digest, bytes=bytes(data))
            return digest

    def get_chunk(self, digest: str) -> Chunk | None:
        with self._lock:
            chunk = self._chunks.get(digest)
            if chunk is None:
                return None
            return Chunk(digest=chunk.digest, bytes=chunk.bytes, stored=chunk.stored)

    def pin_manifest(self, name: str, digests: list[str]) -> Manifest:
        """Pin a manifest. Every chunk must already exist."""
        with self._lock:
            total = 0
            for d in digests:
                chunk = self._chunks.get(d)
                if chunk is None:
                    raise LafsError(f"unknown chunk {d}")
                total += len(chunk.bytes)
            manifest = Manifest(
                name=name,
                chunks=list(digests),
                total_bytes=total,
            )
            self._manifests[name] = manifest
            return Manifest(
                name=manifest.name,
                chunks=list(manifest.chunks),
                total_bytes=manifest.total_bytes,
                id=manifest.id,
                created=manifest.created,
            )

    def read_manifest(self, name: str) -> bytes | None:
        """Reassemble a file through its manifest, verifying every digest."""
        with self._lock:
            manifest = self._manifests.get(name)
            if manifest is None:
                return None
            out = bytearray()
            for d in manifest.chunks:
                chunk = self._chunks.get(d)
                if chunk is None:
                    return None
                if _sha256_hex(chunk.bytes) != d:
                    return None  # corruption — the room's contents changed
                out.extend(chunk.bytes)
            return bytes(out)

    def manifests(self) -> list[Manifest]:
        with self._lock:
            return [
                Manifest(
                    name=m.name,
                    chunks=list(m.chunks),
                    total_bytes=m.total_bytes,
                    id=m.id,
                    created=m.created,
                )
                for m in self._manifests.values()
            ]

    def stats(self) -> dict[str, Any]:
        with self._lock:
            byte_sum = sum(len(c.bytes) for c in self._chunks.values())
            return {
                "chunks": len(self._chunks),
                "bytes": byte_sum,
                "manifests": len(self._manifests),
            }
